R/LearnerSurvCTree.R

```diff
@@ -0,0 +1,115 @@
+            tags = "train"),
+          ParamDbl$new("tol", lower = 0, tags = "train"),
+          ParamInt$new("maxsurrogate", lower = 0L, default = 0L,
+            tags = "train"),
+          ParamLgl$new("numsurrogate", default = FALSE, tags = "train"),
+          ParamInt$new("mtry", lower = 0L, special_vals = list(Inf),
+            default = Inf, tags = "train"),
+          ParamInt$new("maxdepth", lower = 0L, special_vals = list(Inf),
+            default = Inf, tags = "train"),
+          ParamLgl$new("multiway", default = FALSE, tags = "train"),
+          ParamInt$new("splittry", lower = 0L, default = 2L, tags = "train"),
+          ParamLgl$new("intersplit", default = FALSE, tags = "train"),
+          ParamLgl$new("majority", default = FALSE, tags = "train"),
+          ParamLgl$new("caseweights", default = FALSE, tags = "train"),
+          ParamUty$new("applyfun", tags = "train"),
+          ParamInt$new("cores", special_vals = list(NULL), default = NULL,
+            tags = "train"),
+          ParamLgl$new("saveinfo", default = TRUE, tags = "train"),
+          ParamLgl$new("update", default = FALSE, tags = "train"),
+          ParamFct$new("splitflavour", default = "ctree",
+            levels = c("ctree", "exhaustive"), tags = c("train", "control")),
+          ParamUty$new("offset", tags = "train"),
+          ParamUty$new("cluster", tags = "train"),
+          ParamUty$new("scores", tags = "train"),
+          ParamLgl$new("doFit", default = TRUE, tags = "train"),
+          ParamUty$new("pargs", default = mvtnorm::GenzBretz, tags = "train")
+        )
+      )
+      ps$add_dep("nresample", "testtype", CondEqual$new("MonteCarlo"))
+
+      super$initialize(
+        id            = "surv.ctree",
+        packages      = c("partykit", "coin", "distr6"),
+        feature_types = c("integer", "numeric", "factor", "ordered"),
+        predict_types = c("distr", "crank"),
+        param_set     = ps,
+        properties    = "weights"
+      )
+    }
+  ),
+
+  private = list(
+    .train = function(task) {
+      pars = self$param_set$get_values(tags = "train")
+
+      if ("weights" %in% task$properties) {
+        pars$weights = task$weights$weight
+      }
+
+      mod <- mlr3misc::invoke(partykit::ctree, formula = task$formula(),
+        data = task$data(), .args = pars)
+    },
+
+    .predict = function(task) {
+      newdata = task$data(cols = task$feature_names)
+
+      p = mlr3misc::invoke(predict, self$model, type = "prob", newdata = newdata)
+
+      # Define WeightedDiscrete distr6 distribution from the survival function
+      x = lapply(p, function(z) data.frame(x = z$time, cdf = 1 - z$surv))
+      distr = distr6::VectorDistribution$new(
+        distribution = "WeightedDiscrete",
+        params = x,
+        decorators = c("CoreStatistics", "ExoticStatistics"))
+
+      # Define crank as the mean of the survival distribution
+      crank = as.numeric(sum(x[[1]][,1] * c(x[[1]][,2][1], diff(x[[1]][,2]))))
```

Comment:
As the model directly returns median survival time, I think it could make sense to instead set `crank` to be this median survival time, just make sure it's clear in the documentation either way

Reply:
could do, but the predicted median survival time will often be `Inf` if the predicted/estimated survival curve doesn't go below .5

Reply:
argh, this is another instance of "improper" distributions appearing.
We haven´t thought settled yet where/how to handle them, right, @RaphaelS1 ?

Reply:
though, for crank, isn´t it fine to return inf in this case?

Reply:
no Adi is correct, this is indeed what we suggested on as this is such a common case

Reply:
confused - I wasn´t contradicting @adibender here, I was just saying that we do not seem to have the case of improper survival functions covered in the distr interface. Which is a more general problem than "median doesn´t exist".